Add "Memory consistency models" subsection to "Memory orderings" section

yutingshih opened the pull request:
This pull request aims to enhance the comprehensiveness of the “Memory orderings” section of this book. Currently, the section primarily covers memory orders defined in C11/C++11 but lacks discussion on hardware assumptions regarding memory consistency. Thus I made the following changes:

1. Reorganization
    - Moved the content originally under “Memory orderings” related to C11/C++11 atomics to a dedicated subsection titled “C11/C++ atomics” within the “Memory orderings” section.
2. Addition
    - Introduced a new subsection titled “Memory consistency models”
        - Detailed descriptions of sequential consistency, x86 total store order, and ARM’s relaxed memory order.
        - Included schematic diagrams illustrating hardware architectures and litmus tests to aid readers’ comprehension.

yutingshih committed:
Add new subsection "Memory consistency model"

yutingshih committed:
Add weaker models (x86-TSO and ARM) to the "Memory consistency" section

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,141 @@ \section{Do we always need sequentially consistent operations?}
 
+\subsection{Memory consistency models}
+
+When a program is compiled and executed, it doesn't always follow the written order. The system may change the sequence and optimize it to simulate line-by-line execution, as long as the final result matches the expected outcome.
```
Split the sentence starting with "The system may change .." into a new line.

jserv replied:
Follow the same manner for newlines.

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,141 @@ \section{Do we always need sequentially consistent operations?}
+y = 1;             r2 = x;
+\end{ccode}
+
+If this program satisfies sequential consistency, then for Thread 1, \monobox{x = 1} must occur before \monobox{y = 1}, and for Thread 2, \monobox{r1 = y} must occur before \monobox{r2 = x}. For the entire program, the following six execution orders are possible:
```
Use rich LaTeX annotations to represent the following.

yutingshih replied:
Did you mean I should use `\cc{...}` or `\cpp{...}` instead of `\monobox{...}`?

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,141 @@ \section{Do we always need sequentially consistent operations?}
+
+\subsubsection{Total store order (TSO)}
+
+Although sequential consistency is considered the "golden standard" for multi-threaded programs, its many constraints limit performance optimization. As a result, it is rarely implemented in modern processors. Instead, more relaxed memory models are used, such as the total store order (TSO) memory model adopted by the x86 architecture. One can envision the hardware roughly as follows:
```
Change the quotation mark. See https://www.maths.tcd.ie/~dwilkins/LaTeXPrimer/QuotDash.html

yutingshih committed:
Fix inconsistency of code style

1. split setences into separate lines
2. change the quotation marks "..." as ``...''

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,178 @@ \section{Do we always need sequentially consistent operations?}
+If this program satisfies sequential consistency, then for Thread 1, \monobox{x = 1} must occur before \monobox{y = 1}, and for Thread 2, \monobox{r1 = y} must occur before \monobox{r2 = x}.
+For the entire program, the following six execution orders are possible:
+
+\begin{verbatim}
```
Replace the following with table or something alike in LaTeX.

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,178 @@ \section{Do we always need sequentially consistent operations?}
+
+Once Thread 3 reads \monobox{r1 = 1}, \monobox{r2 = 0}, it indicates that the write \monobox{x = 1} reached shared memory before \monobox{y = 1}.
+If at this point Thread 4 reads \monobox{r3 = 1}, it means both writes \monobox{y = 1} and \monobox{x = 1} are visible to Thread 4, so \monobox{r4} can only be \monobox{1}.
+We can say "Thread 1's write to \monobox{x}" happens before "Thread 2's write to \monobox{y}".
```
Be aware of the quotation marks.
See https://www.maths.tcd.ie/~dwilkins/LaTeXPrimer/QuotDash.html

yutingshih committed:
Replace the execution results of code with LaTeX table syntax

yutingshih committed:
Fix the inconsistency of quotation marks

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,209 @@ \section{Do we always need sequentially consistent operations?}
 
+\subsection{Memory consistency models}
+
+When a program is compiled and executed, it doesn't always follow the written order.
```
Don't use short form. i.e., use "does not" instead.

jserv replied:
?

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,209 @@ \section{Do we always need sequentially consistent operations?}
+\subsection{Memory consistency models}
+
+When a program is compiled and executed, it doesn't always follow the written order.
+The system may change the sequence and optimize it to simulate line-by-line execution, as long as the final result matches the expected outcome.
```
You should address the impact from modern microprocessors and optimizing compilers.

jserv commented on concurrency-primer.tex:
```diff
@@ -747,6 +747,209 @@ \section{Do we always need sequentially consistent operations?}
+When a program is compiled and executed, it doesn't always follow the written order.
+The system may change the sequence and optimize it to simulate line-by-line execution, as long as the final result matches the expected outcome.
+
+This requires an agreement between the programmer and the system (hardware, compiler, etc.), ensuring that if the rules are followed, the execution will be correct.
```
Merge the statement into the previous paragraph.

jserv replied:
?

jserv reviewed (changes requested):
Rebase the latest `main` branch.

yutingshih committed:
Address text consistency and merge related statements

- Changed "doesn't" to "does not" for formal English writing
- Fixed the inconsistency of quotation marks
- Merged related statements into one paragraph

jserv reviewed (changes requested):
Instead of `git merge`, do `git rebase`.

jserv commented on concurrency-primer.tex:
```diff
@@ -1090,6 +1090,208 @@ \section{Do we always need sequentially consistent operations?}
+
+The write order seen by one thread can differ from the order seen by other threads because write operations can be reordered during propagation.
+However, reads and writes to the same memory address must still follow a total order.
+Therefore, the following litmus test cannot result in \monobox{r1 = 1}, \monobox{r2 = 2}, but \monobox{r3 = 2}, \monobox{r4 = 1}.
```
Provide some information about running Litmus tests.

jserv commented on concurrency-primer.tex:
```diff
@@ -1090,6 +1090,208 @@ \section{Do we always need sequentially consistent operations?}
+
+\subsubsection{Sequential consistency (SC)}
+
+In the 1970s, Leslie Lamport proposed the most common memory consistency model, sequential consistency (SC), defined as follows:
```
Add some references at the end of this document.

jserv commented:
Any progress?

yutingshih committed:
Merge branch 'sysprog21:main' into main

jserv commented:
Close in favor of #19
